controllers/istiorevisiontag/istiorevisiontag_controller.go

```diff
@@ -0,0 +1,522 @@
+	if reconciledCondition.Status != metav1.ConditionTrue {
+		return reconciledCondition.Reason
+	}
+	if inUseCondition.Status != metav1.ConditionTrue {
+		return inUseCondition.Reason
+	}
+	return v1alpha1.IstioRevisionTagReasonHealthy
+}
+
+func (r *Reconciler) determineReconciledCondition(err error) v1alpha1.IstioRevisionTagCondition {
+	c := v1alpha1.IstioRevisionTagCondition{Type: v1alpha1.IstioRevisionTagConditionReconciled}
+
+	if err == nil {
+		c.Status = metav1.ConditionTrue
+	} else {
+		c.Status = metav1.ConditionFalse
+		c.Reason = v1alpha1.IstioRevisionTagReasonReconcileError
+		c.Message = fmt.Sprintf("error reconciling resource: %v", err)
+	}
+	return c
+}
+
+func (r *Reconciler) determineInUseCondition(ctx context.Context, tag *v1alpha1.IstioRevisionTag) (v1alpha1.IstioRevisionTagCondition, error) {
+	c := v1alpha1.IstioRevisionTagCondition{Type: v1alpha1.IstioRevisionTagConditionInUse}
+
+	isReferenced, err := r.isRevisionTagReferencedByWorkloads(ctx, tag)
+	if err == nil {
+		if isReferenced {
+			c.Status = metav1.ConditionTrue
+			c.Reason = v1alpha1.IstioRevisionTagReasonReferencedByWorkloads
+			c.Message = "Referenced by at least one pod or namespace"
+		} else {
+			c.Status = metav1.ConditionFalse
+			c.Reason = v1alpha1.IstioRevisionTagReasonNotReferenced
+			c.Message = "Not referenced by any pod or namespace"
+		}
+		return c, nil
+	}
+	c.Status = metav1.ConditionUnknown
+	c.Reason = v1alpha1.IstioRevisionTagReasonUsageCheckFailed
+	c.Message = fmt.Sprintf("failed to determine if revision tag is in use: %v", err)
+	return c, fmt.Errorf("failed to determine if IstioRevisionTag is in use: %w", err)
+}
+
+func (r *Reconciler) updateLabels(ctx context.Context, tag *v1alpha1.IstioRevisionTag, rev *v1alpha1.IstioRevision) error {
+	updatedTag := tag.DeepCopy()
+	if rev == nil {
+		delete(updatedTag.Labels, sailOperatorReferencedRevisionLabel)
+	} else {
+		if updatedTag.Labels == nil {
+			updatedTag.Labels = make(map[string]string, 1)
+		}
+		updatedTag.Labels[sailOperatorReferencedRevisionLabel] = rev.Name
+	}
+	return r.Patch(ctx, updatedTag, client.MergeFrom(tag))
+}
+
+func (r *Reconciler) isRevisionTagReferencedByWorkloads(ctx context.Context, tag *v1alpha1.IstioRevisionTag) (bool, error) {
+	log := logf.FromContext(ctx)
+	nsList := corev1.NamespaceList{}
+	nsMap := map[string]corev1.Namespace{}
+	if err := r.Client.List(ctx, &nsList); err != nil { // TODO: can we optimize this by specifying a label selector
+		return false, fmt.Errorf("failed to list namespaces: %w", err)
+	}
+	for _, ns := range nsList.Items {
+		if namespaceReferencesRevisionTag(ns, tag) {
+			log.V(2).Info("Revision is referenced by Namespace", "Namespace", ns.Name)
+			return true, nil
+		}
+		nsMap[ns.Name] = ns
+	}
+
+	podList := corev1.PodList{}
+	if err := r.Client.List(ctx, &podList); err != nil { // TODO: can we optimize this by specifying a label selector
+		return false, fmt.Errorf("failed to list pods: %w", err)
+	}
```

Comment:
I wonder if this controller should be checking for pods at all or if it can defer to the `istiorevision` controller for that which is already checking for the pod annotation that is independent of tag? Basically if this controller's istiorev ref is in use then this tag should be considered in use?

Reply:
Main question I have is this ^^. Rest is basically nits.

Reply:
hmm. yes that is an optimization, at least for the pods. for the namespaces we'd still need to check